[Dash] Support ttml subs without codecs

matthuisman opened the pull request:
Allow the same behaviour as wvtt by setting the codec to ttml for mimeType=application/ttml+xml
This also adds support for mimeType="text/vtt" on representations (not adaptionset) to work correctly
Also fixes a bug where using codecs="wvtt" without a mimeType, SUBTITLESTREAM would not be set.

matthuisman committed:
support ttml subs without codecs

matthuisman commented:
getting odd error
"undefined reference to `adaptive::AdaptiveTree::Representation::SUBTITLESTREAM'"

But i'm using
adaptive::DASHTree::Representation::SUBTITLESTREAM

works fine on windows..

matthuisman commented on src/parser/DASHTree.cpp:
```diff
@@ -710,7 +710,8 @@ static void XMLCALL start(void* data, const char* el, const char** attr)
                 else if (strncmp(dash->current_adaptationset_->mimeType_.c_str(), "application",
-                                 11) == 0)
+                                 11) == 0 ||
+                         strncmp(dash->current_adaptationset_->mimeType_.c_str(), "text", 4) == 0)
```
this is same as existing adaption set code below but for representation attributes
https://github.com/xbmc/inputstream.adaptive/blob/Matrix/src/parser/DASHTree.cpp#L931

without this mimeType="text/vtt" on the **representation** won't work but mimeType="application/ttml+xml" would work

glennguy commented:
Try `#include "../common/AdaptiveTree.h"` in TestHelper.h

edit: nvm, should already be included in DASH/HLSTree.h

glennguy commented:
Due to `ATTRIBUTE_HIDDEN` we can't access this way.

Use `DASHTestTree::SUBTITLE` and `DASHTestTree::Representation::SUBTITLESTREAM`

edit: that doesn't fix the issue

glennguy commented:
I think the correct solution is to assign `adaptive::AdaptiveTree::Representation::SUBTITLESTREAM` to a variable outside of the test assertion (`EXPECT_EQ`), then we will avoid the visibility issue.

matthuisman commented:
@glennguy 
Success! thank you.
Added a comment in the code as well to remind us.

also added more tests because - why not!

ready for review and merge :)

glennguy commented on src/parser/DASHTree.cpp:
```diff
@@ -722,6 +723,28 @@ static void XMLCALL start(void* data, const char* el, const char** attr)
+              dash->current_representation_->codecs_.empty())
+            dash->current_representation_->codecs_ = "wvtt";
+
+          if (dash->current_adaptationset_->mimeType_ == "application/ttml+xml" &&
```
`else if` might be better here

glennguy commented on src/parser/DASHTree.cpp:
```diff
@@ -722,6 +723,28 @@ static void XMLCALL start(void* data, const char* el, const char** attr)
+            dash->current_adaptationset_->type_ = DASHTree::SUBTITLE;
+            dash->current_adaptationset_->mimeType_ = "application/ttml+xml";
+          }
+
```
These could both be nested under 1 condition for if (dash->current_adaptationset_->type_ != DASHTree::SUBTITLE)

matthuisman replied:
how so?
they both need to check for text/vtt or ttml
so if I did just a check for ! SUBTITLE, I'd need two more checks inside that to check the codec.

glennguy commented:
Looks great, going in :)
